src/Orleans/AssemblyLoader/AssemblyLoader.cs

```diff
@@ -361,40 +358,71 @@ private bool ReflectionOnlyLoadAssembly(string pathName, out Assembly assembly,
+
+                            var isManaged = isPureIL || processorArchitecture == ProcessorArchitecture.MSIL;
 
-            if (peKind.HasFlag(PortableExecutableKinds.NotAPortableExecutableImage) ||
-                peKind.HasFlag(PortableExecutableKinds.Unmanaged32Bit))
+
+                            var isLoadable = (isPureIL && processorArchitecture == ProcessorArchitecture.MSIL) ||
+                                                 (Environment.Is64BitProcess && processorArchitecture == ProcessorArchitecture.Amd64) ||
+                                                 (!Environment.Is64BitProcess && processorArchitecture == ProcessorArchitecture.X86);
+
+                            if (!isLoadable)
+                            {
+                                complaints = new[] { $"The file {fileName} is not loadable into this process, either it is not an MSIL assembly or the compliled for a different processor architecture." };
+                            }
+
+                            return isLoadable;
+                        }
+                        else
+                        {
+                            complaints = new[] { $"The file {fileName} does not contain any CLR metadata, probably it is a native file." };
+                            return false;
+                        }
+                    }
+                }
+            }
+            catch (IOException ex)
             {
-                // this block of code should never run since the assembly was successfully loaded
-                throw new InvalidOperationException("Unexpected block of code reached");
+                return false;
             }
-
-            if ((peKind.HasFlag(PortableExecutableKinds.Required32Bit) && Environment.Is64BitProcess) || 
-                (peKind.HasFlag(PortableExecutableKinds.PE32Plus) && !Environment.Is64BitProcess))
+            catch (BadImageFormatException ex)
             {
-                // targets wrong bitness
-                complaints = new[] { $"The assembly {assembly.FullName} is compiled for a different platform than the running process" };
                 return false;
             }
-
-            if (machine != this.exeInfo.Value.MachineType)
+            catch (UnauthorizedAccessException ex)
             {
-                complaints = new[] { $"The assembly {assembly.FullName} was compiled for {machine} but the current exe was compiled for {this.exeInfo.Value.MachineType}" };
                 return false;
             }
-
-            return true;
+            catch (MissingMethodException ex)
+            {
+                complaints = new[] { "MissingMethodException occured. Please try to add a BindingRedirect for System.Collections.ImmutableCollections to the App.config file to correct this error." };
+                return false;
+            }
+            catch (Exception ex)
+            {
+                complaints = new[] { ex.ToString() };
```

Comment:
Are we doing that other places? this was not that, I can add of course, but complaints are printed otherwise I think.